bindings/matrix-sdk-ffi/src/event.rs

```diff
@@ -0,0 +1,183 @@
+        self.0.sender().to_string()
+    }
+
+    pub fn event_type(&self) -> Result<TimelineEventType, ClientError> {
+        let event_type = match &self.0 {
+            AnySyncTimelineEvent::MessageLike(event) => {
+                TimelineEventType::MessageLike { content: event.clone().try_into()? }
+            }
+            AnySyncTimelineEvent::State(event) => {
+                TimelineEventType::State { content: event.clone().try_into()? }
+            }
+        };
+        Ok(event_type)
+    }
+}
+
+#[derive(uniffi::Enum)]
+pub enum TimelineEventType {
+    MessageLike { content: MessageLikeEventContent },
+    // we don't support state events yet so they do not have any content
+    State { content: StateEventContent },
+}
+
+#[derive(uniffi::Enum)]
+pub enum StateEventContent {
+    PolicyRuleRoom,
+    PolicyRuleServer,
+    PolicyRuleUser,
+    RoomAliases,
+    RoomAvatar,
+    RoomCanonicalAlias,
+    RoomCreate,
+    RoomEncryption,
+    RoomGuestAccess,
+    RoomHistoryVisibility,
+    RoomJoinRules,
+    RoomMember { user_id: String, membership_state: MembershipState },
+    RoomName,
+    RoomPinnedEvents,
+    RoomPowerLevels,
+    RoomServerAcl,
+    RoomThirdPartyInvite,
+    RoomTombstone,
+    RoomTopic,
+    SpaceChild,
+    SpaceParent,
+}
+
+impl TryFrom<AnySyncStateEvent> for StateEventContent {
+    type Error = anyhow::Error;
+
+    fn try_from(value: AnySyncStateEvent) -> anyhow::Result<Self> {
+        let event = match value {
+            AnySyncStateEvent::PolicyRuleRoom(_) => StateEventContent::PolicyRuleRoom,
+            AnySyncStateEvent::PolicyRuleServer(_) => StateEventContent::PolicyRuleServer,
+            AnySyncStateEvent::PolicyRuleUser(_) => StateEventContent::PolicyRuleUser,
+            AnySyncStateEvent::RoomAliases(_) => StateEventContent::RoomAliases,
+            AnySyncStateEvent::RoomAvatar(_) => StateEventContent::RoomAvatar,
+            AnySyncStateEvent::RoomCanonicalAlias(_) => StateEventContent::RoomCanonicalAlias,
+            AnySyncStateEvent::RoomCreate(_) => StateEventContent::RoomCreate,
+            AnySyncStateEvent::RoomEncryption(_) => StateEventContent::RoomEncryption,
+            AnySyncStateEvent::RoomGuestAccess(_) => StateEventContent::RoomGuestAccess,
+            AnySyncStateEvent::RoomHistoryVisibility(_) => StateEventContent::RoomHistoryVisibility,
+            AnySyncStateEvent::RoomJoinRules(_) => StateEventContent::RoomJoinRules,
+            AnySyncStateEvent::RoomMember(content) => {
+                let state_key = content.state_key().to_string();
+                let original_content = content
+                    .as_original()
+                    .context("Failed to get original content")?
+                    .content
+                    .clone();
+                StateEventContent::RoomMember {
+                    user_id: state_key,
+                    membership_state: original_content.membership.into(),
+                }
```

Comment:
@jplatte 
Had to implement this quite a lot, but I am unable to implement a generic function that does this automatically, I get this error on this implementation:
```
fn try_convert<T: StaticStateEventContent + RedactContent>(
    event: ruma::events::SyncStateEvent<T>,
) -> anyhow::Result<T>
where
    T::Redacted: RedactedStateEventContent,
{
    let original_content =
        event.as_original().context("Failed to get original content")?.content.clone();
    Ok(original_content)
}
```

the method `as_original` exists for enum `SyncStateEvent<T>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`<<T as ruma::events::RedactContent>::Redacted as ruma::events::RedactedStateEventContent>::StateKey = <T as ruma::events::StateEventContent>::StateKey`